### ISSUE TITLE
win_pkg: package removal ignores cache_dir parameter

### ISSUE THREAD

**Issue by morganwillcock** [saltstack/salt#33119]
Because cache_dir isn't used for package removal, an uninstaller that needs a config file must rely on the install process having cached it, and the cache remaining intact. 

As an example for installing and uninstalling Microsoft Office:

> installer: 'salt://win/repo-ng/microsoft_office_professional_plus_2013/setup.exe'
> install_flags: '/adminfile full.msp'
> uninstaller: 'salt://win/repo-ng/microsoft_office_professional_plus_2013/setup.exe'
> uninstall_flags: '/uninstall ProPlus /config SilentUninstall.xml'
> cache_dir: True

Uninstallation will fail if 'SilentUninstall.xml' is not present. Since the directory is not cached again, this file may or may not exist when the uninstall process is run.

Tested on 2015.8.8-2.

**Comment by Ch3LL** [comment]
@morganwillcock I think this is a good approach to ensure we cache these files to ensure an uninstall is available. 

@twangboy do you think this is the right approach and is possible?

**Comment by damon-atkins** [comment]
I would of thought that part of a state file would be  deployment of the configuration file and then pkg.install (or remove) would be dependant on it.

Because the git  repo (pkg's definitions) needs to be generic, if something is more specific it would be copied and called something else.

**Comment by morganwillcock** [comment]
It can't depend on a state because that means it can't be run on-demand using the execution module (would seem to render win_pkg redundant), plus cache_path isn't available as an environment variable or a grain so you are going to end up hard coding matching file paths in the state and the package definition. There is also the security issue of running unverified files or configurations as SYSTEM when you run an uninstaller, if you aren't actively checking the hash beforehand.

I'm not using the git repos, SilentUninstall.xml would be rendered or copied via syndic.

**Comment by Ch3LL** [comment]
I spoke with @twangboy and we decided this would be an appropriate feature request, although he needs to think through on how to approach this. I'll label it a feature request and he has assigned it to himself to find a solution to this request. Thanks

**Comment by hrumph** [comment]
Just some commentary. To deal with office 2013, I did the following:
```
ms office 2013 config file present:
    file.managed:
        - name: "C:\\Program Files (x86)\\Common Files\\Microsoft Shared\\OFFICE15\\Office Setup Controller\\config.x
ml"
        - makedirs: True
        - source: salt://win/ms-office/2013_x86/config.xml


ms office 2013 not installed:
    pkg.removed:
        - name: ms-office_x86
        - refresh: False
        - require:
             - file: ms office 2013 config file present
```

Note that the the installers are automatically put on the C: drive for you. My repo ng files look like this:


```
ms-office:
    '16.0.4266.1001':
         installer: salt://win/ms-office/2016/setup.exe
         full_name: Microsoft Office Professional Plus 2016
         reboot: False
         cache_dir: True
         install_flags: '/adminfile mike_config.msp'
         uninstaller: 'C:\Program Files\Common Files\Microsoft Shared\OFFICE16\Office Setup Controller\setup.exe'
         uninstall_flags: '/uninstall PROPLUS /dll OSETUP.DLL'
```
and
```

```
ms-office_x86:
    '15.0.4569.1506':
         installer: salt://win/ms-office/2013_x86/setup.exe
         full_name: Microsoft Office Professional Plus 2013
         reboot: False
         cache_dir: True
         install_flags: '/adminfile mike_config.MSP'
         uninstaller: 'C:\Program Files (x86)\Common Files\Microsoft Shared\OFFICE15\Office Setup Controller\setup.ex
e'
         uninstall_flags: '/uninstall PROPLUS /dll OSETUP.DLL'
```

And insofar as I can remember, for ms 2016, if you place the config.xml file in the installer dir, the Office 2016 installer will automatically copy it over to C:\Program Files\Common Files\Microsoft Shared\OFFICE16\Office Setup Controller\. In otherwords, as I see it, salt does _not_ have to manage this file for Office 2016.

I wonder if this is really a non-issue because if an installer doesn't install with everything you need to uninstall with then, it could be argued, it's the fault of the installer, and it may not be a good idea to overcomplicate things in order to bandage over other thing's faults.

**Comment by morganwillcock** [comment]
Thanks for the info. For my use case I'm not the only one making changes to the system, hence using state based software rather than just remote management tools. Unfortunately I can't guarantee the integrity of anything on the disk, so the hash check that is part of the transfer (for programs I'm about to run as the system account) are, for me at least, a big deal.

**Comment by hrumph** [comment]
Well you phrased this issue as a sort of bug, buy to my mind it's definitely a feature request.

**Comment by damon-atkins** [comment]
msi are cached under c:\Windows\Installer   However msi also look after themselves.  Other installers should install an uninstaller with the program.

**Comment by morganwillcock** [comment]
@hrumph from a security perspective I would say it's bug, as it's not checking that any referenced file is what you'd expect it to be (ideally I would like to supply a dictionary that maps to required install or uninstall data, which would also allow you to use cache_dir with web servers).

@damon-atkins Yes. I work in education so unfortunately a lot of the software has home-brew installers (MSI packages are pretty rare) or just isn't tested for silent install or uninstall. There are quite a lot if instances where the uninstall information is in the wrong place, e.g. anything built with WiX that offers per-user or per-machine installation will put the uninstall data in the user profile of the installing user (even if you do a per-machine installation).

**Comment by damon-atkins** [comment]
Hi @morganwillcock, the current version does not support per user installs. I don't think it will even list it as being installed. The version I am working on does list software which is installed under a user.

Salt Pkg does not support GUI uninstalls/installs.   See also www.autoitscript.com & pywinauto 

I am hopping to get back into it in a few weeks.

**Comment by hrumph** [comment]
@morganwillcock, I don't really understand your situation. Supposing that users have admin rights (like they often do) and overwrite the uninstallers with something else, then so what? They can do anything they please already, because they have admin rights. If they don't have admin rights then they won't be able to overwrite anything.

**Comment by Ch3LL** [comment]
This has been fixed in #40230. Closing issue. Please let me know if this fix does not work and we can re-open

**Comment by morganwillcock** [comment]
@Ch3LL thank you, I'll give it a try
